File: libs/intersector.php
<?php
//sources:
//  https://stackoverflow.com/questions/2255842/detecting-coincident-subset-of-two-coincident-line-segments/2255848#2255848
//https://stackoverflow.com/questions/1073336/circle-line-segment-collision-detection-algorithm

// port of this JavaScript code with some changes:
//   http://www.kevlindev.com/gui/math/intersection/Intersection.js
// found here:
//   http://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect/563240#563240

//require_once('vertex.php');
require_once(__DIR__ . DIRECTORY_SEPARATOR . 'vertex.php');

class Intersector {
    /** @var float */
    static $MyEpsilon = 0.00001;

    private static function OverlapIntervals($ub1, $ub2): array {
        $l = min($ub1, $ub2);
        $r = max($ub1, $ub2);
        $A = max(0, $l);
        $B = min(1, $r);
        if ($A > $B) {// no intersection
            return [];
        } else if ($A == $B) {
            return [$A];
        } else { // if (A < B)
            return [$A, $B];
        }
    }

    // IMPORTANT: $a1 and $a2 cannot be the same, e.g. $a1--$a2 is a true segment, not a point
    // $b1/$b2 may be the same ($b1--$b2 is a point)
    /**
     * @param Vertex $a1
     * @param Vertex $a2
     * @param Vertex $b1
     * @param Vertex $b2
     * @return array
     */
    private static function oneD_Intersection(Vertex $a1, Vertex $a2, Vertex $b1, Vertex $b2): array {
        //float ua1 = 0.0f; // by definition
        //float ua2 = 1.0f; // by definition
        /** @var float $ub1 */
        /** @var float $ub2 */
        //$ub1 = 0; $ub2 = 0;

        /** @var float $denomx */
        /** @var float $denomy */
        $denomx = $a2->X() - $a1->X();
        $denomy = $a2->Y() - $a1->Y();

        if (abs($denomx) > abs($denomy)) {
            $ub1 = ($b1->X() - $a1->X()) / $denomx;
            $ub2 = ($b2->X() - $a1->X()) / $denomx;
        } else {
            $ub1 = ($b1->Y() - $a1->Y()) / $denomy;
            $ub2 = ($b2->Y() - $a1->Y()) / $denomy;
        }

        $ret = [];//new List<Vertex>();

        /** @var array $interval */
        $interval = Intersector::OverlapIntervals($ub1, $ub2);
        /** @var float $f */
        foreach ($interval as $f) {
            /** @var float $x */
            /** @var float $y */
            $x = $a2->X() * $f + $a1->X() * (1.0 - $f);
            $y = $a2->Y() * $f + $a1->Y() * (1.0 - $f);
            /** @var Vertex $p */
            $p = new Vertex($x, $y);
            $ret[] = $p;
        }
        return $ret;
    }

    /**
     * @param Vertex $p
     * @param Vertex $a1
     * @param Vertex $a2
     * @return bool
     * @throws Exception
     */
    private static function pointOnLine(Vertex $p, Vertex $a1, Vertex $a2): bool {
        /** @var float $dummyU */
        $dummyU = 0.0;
        $d = Intersector::distFromSeg($p, $a1, $a2, Intersector::$MyEpsilon, $dummyU);
        return $d < Intersector::$MyEpsilon
            && ($p->x >= $a1->x && $p->x <= $a2->x || $p->x >= $a2->x && $p->x <= $a1->x)
            && ($p->y >= $a1->y && $p->y <= $a2->y || $p->y >= $a2->y && $p->y <= $a1->y);
    }

    /**
     * @param Vertex $p
     * @param Vertex $q0
     * @param Vertex $q1
     * @param double $radius
     * @param float $u
     * @return double
     * @throws Exception
     */
    private static function distFromSeg(Vertex $p, Vertex $q0, Vertex $q1, float $radius, float &$u): float {
        // formula here:
        //http://mathworld.wolfram.com/Point-LineDistance2-Dimensional.html
        // where x0,y0 = $p
        //       x1,y1 = $q0
        //       x2,y2 = $q1
        /** @var double $dx21 */
        /** @var double $dy21 */
        /** @var double $dx10 */
        /** @var double $dy10 */
        /** @var double $dx21 */
        $dx21 = $q1->X() - $q0->X();
        $dy21 = $q1->Y() - $q0->Y();
        $dx10 = $q0->X() - $p->X();
        $dy10 = $q0->Y() - $p->Y();
        $segLength = sqrt($dx21 * $dx21 + $dy21 * $dy21);
        if ($segLength < Intersector::$MyEpsilon) {
            throw new Exception("Expected line segment, not point.");
        }
        /** @var double $num */
        $num = abs($dx21 * $dy10 - $dx10 * $dy21);
        /** @var double $d */
        $d = $num / $segLength;
        return $d;
    }

    /** this is the general case. Really really general
     * @param Vertex $a1
     * @param Vertex $a2
     * @param Vertex $b1
     * @param Vertex $b2
     * @return array
     * @throws Exception
     */
    public static function intersection(Vertex $a1, Vertex $a2, Vertex $b1, Vertex $b2): array {
        if ($a1->equals($a2) && $b1->equals($b2)) {
            // both "segments" are points, return either point
            if ($a1->equals($b1)) {
                return [$a1];
            } else // both "segments" are different points, return empty set
            {
                return [];
            }
        } else if ($b1->equals($b2)) // b is a point, a is a segment
        {
            if (Intersector::pointOnLine($b1, $a1, $a2)) {
                return [$b1];
            } else {
                return [];
            }
        } else if ($a1->equals($a2)) // a is a point, b is a segment
        {
            if (Intersector::pointOnLine($a1, $b1, $b2)) {
                return [$a1];
            } else {
                return [];
            }
        }

        // at this point we know both a and b are actual segments

        /** @var float $ua_t */
        /** @var float $ub_t */
        /** @var float $u_b */
        $ua_t = ($b2->X() - $b1->X()) * ($a1->Y() - $b1->Y()) - ($b2->Y() - $b1->Y()) * ($a1->X() - $b1->X());
        $ub_t = ($a2->X() - $a1->X()) * ($a1->Y() - $b1->Y()) - ($a2->Y() - $a1->Y()) * ($a1->X() - $b1->X());
        $u_b = ($b2->Y() - $b1->Y()) * ($a2->X() - $a1->X()) - ($b2->X() - $b1->X()) * ($a2->Y() - $a1->Y());

        // Infinite lines intersect somewhere
        if (!(-Intersector::$MyEpsilon < $u_b && $u_b < Intersector::$MyEpsilon)) {  // e.g. $u_b != 0.0
            /** @var float $ua */
            /** @var float $ub */
            $ua = $ua_t / $u_b;
            $ub = $ub_t / $u_b;
            if (0.0 <= $ua && $ua <= 1.0 && 0.0 <= $ub && $ub <= 1.0) {
                // Intersection
                return [
                    new Vertex($a1->X() + $ua * ($a2->X() - $a1->X()),
                        $a1->Y() + $ua * ($a2->Y() - $a1->Y()))
                ];
            } else {
                // No Intersection
                return [];
            }
        } else { // lines (not just segments) are parallel or the same line
            // Coincident
            // find the common overlapping section of the lines
            // first find the distance (squared) from one point ($a1) to each point
            if ((-Intersector::$MyEpsilon < $ua_t && $ua_t < Intersector::$MyEpsilon)
                || (-Intersector::$MyEpsilon < $ub_t && $ub_t < Intersector::$MyEpsilon)) {
                if ($a1->equals($a2)) { // danger!
                    return Intersector::oneD_Intersection($b1, $b2, $a1, $a2);
                } else { // safe
                    return Intersector::oneD_Intersection($a1, $a2, $b1, $b2);
                }
            } else {
                // Parallel
                return [];
            }
        }
    }

    public static function lineCircleIntersection(Vertex $l1, Vertex $l2, Vertex $c, float $radius) {
        // Find the points of intersection.)
        //float $dx, $dy, $A, $B, $C, $det, $t;

        $dx = $l2->x - $l1->x;
        $dy = $l2->y - $l1->y;

        $A = $dx * $dx + $dy * $dy;
        $B = 2 * ($dx * ($l1->x - $c->x) + $dy * ($l1->y - $c->y));
        $C = ($l1->x - $c->x) * ($l1->x - $c->x) +
            ($l1->y - $c->y) * ($l1->y - $c->y) -
            $radius * $radius;

        $det = $B * $B - 4 * $A * $C;
        if (($A <= 0.0000001) || ($det < 0)) {
            // No real solutions.
            return [];
        } else if ($det == 0) {
            // One solution.
            $t = -$B / (2 * $A);
            $intersection1 =
                new Vertex($l1->x + $t * $dx, $l1->y + $t * $dy);
            return [$intersection1];
        } else {
            // Two solutions.
            $t = (float)((-$B + sqrt($det)) / (2 * $A));
            $intersection1 =
                new Vertex($l1->x + $t * $dx, $l1->y + $t * $dy);
            $t = (float)((-$B - sqrt($det)) / (2 * $A));
            $intersection2 =
                new Vertex($l1->x + $t * $dx, $l1->y + $t * $dy);
            return [$intersection1, $intersection2];
        }
    }
    public static function lineArcIntersection(Vertex $l1, Vertex $l2, Vertex $a1, Vertex $a2, $ignoreLineArcTouch): array {
        $xc = $a1->Xc();
        $yc = $a1->Yc();
        $xs = $a1->X();
        $ys = $a1->Y();
        $type = $a1->d();
        $circleIntersections = Intersector::lineCircleIntersection($l1, $l2
            , new Vertex($a1->Xc(), $a1->Yc()), Intersector::dist($xc, $yc, $xs, $ys));
        $touch = count($circleIntersections) == 1
            && !$a1->roughtlyEquals($circleIntersections[0])
            && !$a2->roughtlyEquals($circleIntersections[0])
        ;
        if ($touch && $ignoreLineArcTouch) {
            return [];
        }
        $arcAngle1 = Intersector::angle($xc, $yc, $a1->x, $a1->y);
        $arcAngle2 = Intersector::angle($xc, $yc, $a2->x, $a2->y);
        $result = [];
        if ($type == -1) { // clock-wise
            $t = $arcAngle1;
            $arcAngle1 = $arcAngle2;
            $arcAngle2 = $t;
        }
        foreach ($circleIntersections as $intersection) {
            if ($intersection->roughtlyEquals($a1)) {
                $intersection = $a1;
            }
            if ($intersection->roughtlyEquals($a2)) {
                $intersection = $a2;
            }
            if ($intersection->isInside($l1, $l2)) {
                $intersectionAngle = Intersector::angle($xc, $yc, $intersection->x, $intersection->y);
                if ($arcAngle2 >= $arcAngle1) {
                    if ($arcAngle2 >= $intersectionAngle and $intersectionAngle >= $arcAngle1) {
                        $result[] = $intersection;
                    }
                } else {
                    if ($intersectionAngle <= $arcAngle2 or $intersectionAngle >= $arcAngle1) {
                        $result[] = $intersection;
                    }
                }
            }
        }
        return $result;
    }

    /*
     * * Return the distance between two points
     */
    public static function dist($x1, $y1, $x2, $y2) {
        return sqrt(($x1 - $x2) * ($x1 - $x2) + ($y1 - $y2) * ($y1 - $y2));
    }
    /*
     * * Calculate the angle between 2 points, where Xc,Yc is the center of a circle
     * * and x,y is a point on its circumference. All angles are relative to
     * * the 3 O'Clock position. Result returned in radians
     */

    public static function angle($xc, $yc, $x1, $y1) {
        $d = Intersector::dist($xc, $yc, $x1, $y1); // calc distance between two points
        if ($d != 0) {
            if (asin(($y1 - $yc) / $d) >= 0) {
                $a1 = acos(($x1 - $xc) / $d);
            } else {
                $a1 = 2 * pi() - acos(($x1 - $xc) / $d);
            }
        } else {
            $a1 = 0;
        }
        return $a1;
    }
}

File: java/src/Intersector.java
import java.util.ArrayList;
import java.util.List;

public class Intersector {
    public static final double EPSILON = 0.00001;

    public static double dist(double x1, double y1, double x2, double y2) {
        return Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));
    }

    public static double angle(double xc, double yc, double x1, double y1) {
        double d = dist(xc, yc, x1, y1);
        if (!(isZero(d))) {
            if (Math.asin((y1 - yc) / d) >= 0.0) {
                return Math.acos((x1 - xc) / d);
            } else {
                return 2.0 * Math.PI - Math.acos((x1 - xc) / d);
            }
        } else {
            return 0.0;
        }
    }

    private static boolean isZero(double v) {
        return v == 0.0;
    }

    private static boolean isEqual(double a, double b) {
        return a == b;
    }

    /// Line-line segment intersection
    public static List<Vertex> intersection(Vertex a1, Vertex a2, Vertex b1, Vertex b2) {
        List<Vertex> result = new ArrayList<>();

        // Both points
        if (a1.equals(a2) && b1.equals(b2)) {
            if (a1.equals(b1)) result.add(a1.copy());
            return result;
        }
        // b is a point
        if (b1.equals(b2)) {
            if (pointOnLine(b1, a1, a2)) result.add(b1.copy());
            return result;
        }
        // a is a point
        if (a1.equals(a2)) {
            if (pointOnLine(a1, b1, b2)) result.add(a1.copy());
            return result;
        }

        double uaT = (b2.x - b1.x) * (a1.y - b1.y) - (b2.y - b1.y) * (a1.x - b1.x);
        double ubT = (a2.x - a1.x) * (a1.y - b1.y) - (a2.y - a1.y) * (a1.x - b1.x);
        double uB = (b2.y - b1.y) * (a2.x - a1.x) - (b2.x - b1.x) * (a2.y - a1.y);

        if (!(-EPSILON < uB && uB < EPSILON)) {
            double ua = uaT / uB;
            double ub = ubT / uB;
            if (ua >= 0.0 && ua <= 1.0 && ub >= 0.0 && ub <= 1.0) {
                result.add(new Vertex(
                    a1.x + ua * (a2.x - a1.x),
                    a1.y + ua * (a2.y - a1.y)
                ));
            }
        } else {
            // Parallel or coincident
            if ((-EPSILON < uaT && uaT < EPSILON) || (-EPSILON < ubT && ubT < EPSILON)) {
                if (a1.equals(a2)) {
                    result.addAll(oneDIntersection(b1, b2, a1, a2));
                } else {
                    result.addAll(oneDIntersection(a1, a2, b1, b2));
                }
            }
        }
        return result;
    }

    private static List<Double> overlapIntervals(double ub1, double ub2) {
        List<Double> result = new ArrayList<>();
        double l = Math.min(ub1, ub2);
        double r = Math.max(ub1, ub2);
        double a = Math.max(0.0, l);
        double b = Math.min(1.0, r);
        if (a > b) {
            return result;
        } else if (a == b) {
            result.add(a);
        } else {
            result.add(a);
            result.add(b);
        }
        return result;
    }

    private static List<Vertex> oneDIntersection(Vertex a1, Vertex a2, Vertex b1, Vertex b2) {
        double denomx = a2.x - a1.x;
        double denomy = a2.y - a1.y;
        double ub1, ub2;
        if (Math.abs(denomx) > Math.abs(denomy)) {
            ub1 = (b1.x - a1.x) / denomx;
            ub2 = (b2.x - a1.x) / denomx;
        } else {
            ub1 = (b1.y - a1.y) / denomy;
            ub2 = (b2.y - a1.y) / denomy;
        }

        List<Vertex> result = new ArrayList<>();
        for (double f : overlapIntervals(ub1, ub2)) {
            result.add(new Vertex(a2.x * f + a1.x * (1.0 - f), a2.y * f + a1.y * (1.0 - f)));
        }
        return result;
    }

    private static boolean pointOnLine(Vertex p, Vertex a1, Vertex a2) {
        double d = distFromSeg(p, a1, a2);
        return d < EPSILON
            && (p.x >= Math.min(a1.x, a2.x) && p.x <= Math.max(a1.x, a2.x))
            && (p.y >= Math.min(a1.y, a2.y) && p.y <= Math.max(a1.y, a2.y));
    }

    private static double distFromSeg(Vertex p, Vertex q0, Vertex q1) {
        double dx21 = q1.x - q0.x;
        double dy21 = q1.y - q0.y;
        double dx10 = q0.x - p.x;
        double dy10 = q0.y - p.y;
        double segLength = Math.sqrt(dx21 * dx21 + dy21 * dy21);
        if (segLength < EPSILON) {
            return Double.MAX_VALUE;
        }
        return Math.abs(dx21 * dy10 - dx10 * dy21) / segLength;
    }

    /// Line-circle intersection
    public static List<Vertex> lineCircleIntersection(Vertex l1, Vertex l2, Vertex c, double radius) {
        List<Vertex> result = new ArrayList<>();
        double dx = l2.x - l1.x;
        double dy = l2.y - l1.y;
        double a = dx * dx + dy * dy;
        double b = 2.0 * (dx * (l1.x - c.x) + dy * (l1.y - c.y));
        double cc = (l1.x - c.x) * (l1.x - c.x) + (l1.y - c.y) * (l1.y - c.y) - radius * radius;
        double det = b * b - 4.0 * a * cc;

        if (a <= 0.0000001 || det < 0.0) {
            return result;
        } else if (isZero(det)) {
            double t = -b / (2.0 * a);
            result.add(new Vertex(l1.x + t * dx, l1.y + t * dy));
        } else {
            double t1 = (-b + Math.sqrt(det)) / (2.0 * a);
            double t2 = (-b - Math.sqrt(det)) / (2.0 * a);
            result.add(new Vertex(l1.x + t1 * dx, l1.y + t1 * dy));
            result.add(new Vertex(l1.x + t2 * dx, l1.y + t2 * dy));
        }
        return result;
    }

    /// Line-arc intersection
    public static List<Vertex> lineArcIntersection(
            Vertex l1, Vertex l2,
            Vertex a1, Vertex a2,
            boolean ignoreTouch) {
        double xc = a1.seg.xc;
        double yc = a1.seg.yc;
        double radius = dist(xc, yc, a1.x, a1.y);
        List<Vertex> circleInts = lineCircleIntersection(l1, l2, new Vertex(xc, yc), radius);

        boolean touch = circleInts.size() == 1
            && !a1.roughlyEquals(circleInts.get(0))
            && !a2.roughlyEquals(circleInts.get(0));
        if (touch && ignoreTouch) {
            return new ArrayList<>();
        }

        double arcAngle1 = angle(xc, yc, a1.x, a1.y);
        double arcAngle2 = angle(xc, yc, a2.x, a2.y);
        if (a1.seg.d == -1) {
            double tmp = arcAngle1;
            arcAngle1 = arcAngle2;
            arcAngle2 = tmp;
        }

        List<Vertex> result = new ArrayList<>();
        for (Vertex intV : circleInts) {
            if (intV.roughlyEquals(a1)) {
                intV = a1.copy();
            }
            if (intV.roughlyEquals(a2)) {
                intV = a2.copy();
            }
            if (intV.isInside(l1, l2)) {
                double intAngle = angle(xc, yc, intV.x, intV.y);
                boolean inArc;
                if (arcAngle2 >= arcAngle1) {
                    inArc = intAngle >= arcAngle1 && intAngle <= arcAngle2;
                } else {
                    inArc = intAngle <= arcAngle2 || intAngle >= arcAngle1;
                }
                if (inArc) {
                    result.add(intV);
                }
            }
        }
        return result;
    }
}
